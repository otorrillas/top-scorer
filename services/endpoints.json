{
  "gameList": "http://www.speedrun.com/api/v1/games"
}
