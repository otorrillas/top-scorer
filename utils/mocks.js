export const mockRuns = [
  {
    videos: { links: [{ uri: 'https://video-url' }] },
    players: [{ id: 'playerId' }],
    times: { primary_t: 450 }
  }
];

export const mockGames = {
  '1234': {
    abbreviation: 'test',
    assets: {
      'cover-medium': {
        uri: 'https://cover-medium-url'
      }
    },
    runs: mockRuns
  }
};

export const mockGamesResponse = [
  {
    id: '1234',
    ...mockGames['1234']
  },
  {
    id: '45678',
    ...mockGames['1234']
  }
];
