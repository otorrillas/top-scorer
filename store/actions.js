import 'isomorphic-unfetch';

import _find from 'lodash/find';

import endpoints from '../services/endpoints.json';

export const actionTypes = {
  RESET: 'RESET',
  RESET: 'FAILURE',
  FETCH_GAMES_LIST: 'FETCH_GAMES_LIST',
  FETCH_GAMES_LIST_SUCCESS: 'FETCH_GAMES_LIST_SUCCESS',
  FETCH_RUNS: 'FETCH_RUNS',
  FETCH_RUNS_SUCCESS: 'FETCH_RUNS_SUCCESS'
};

export function reset() {
  return { type: actionTypes.RESET };
}

export function failure() {
  return { type: actionTypes.FAILURE };
}

export function fetchGamesList() {
  return async function(dispatch) {
    await fetch(endpoints.gameList)
      .then(response => response.json())
      .then(json => {
        dispatch(fetchGamesListSuccess(json.data));
      })
      .catch(error => console.log(error));
  };
}

export function fetchGamesListSuccess(data) {
  return {
    type: actionTypes.FETCH_GAMES_LIST_SUCCESS,
    data
  };
}

export function fetchRuns(id) {
  return async function(dispatch, getState) {
    const games = getState().games;

    const targetGame = _find(games, { id });
    const runsUrl = _find(targetGame.links, { rel: 'runs' }).uri;

    await fetch(runsUrl)
      .then(response => response.json())
      .then(json => {
        dispatch(fetchRunsSuccess(id, json.data));
      })
      .catch(error => console.log(error));
  };
}

export function fetchRunsSuccess(id, data) {
  return {
    type: actionTypes.FETCH_RUNS_SUCCESS,
    id,
    data
  };
}
