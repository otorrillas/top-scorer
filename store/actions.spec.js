import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import nock from 'nock';

import * as actions from './actions';
import { mockGames, mockRuns } from 'Utils/mocks';
import endpoints from 'Services/endpoints.json';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('Redux - actions', () => {
  test('export action types', () => {
    expect(actions.actionTypes).toMatchSnapshot();
  });

  test('should create a RESET action', () => {
    expect(actions.reset()).toMatchSnapshot();
  });

  test('should create a FAILURE action', () => {
    expect(actions.failure('test error')).toMatchSnapshot();
  });

  test('should create a FETCH_GAMES_LIST_SUCCESS action', () => {
    expect(actions.fetchGamesListSuccess(mockGames)).toMatchSnapshot();
  });

  test('should create a FETCH_RUN_SUCCESS action', () => {
    expect(actions.fetchRunsSuccess(1234, mockRuns)).toMatchSnapshot();
  });

  describe('FETCH_GAMES_LIST', () => {
    const store = mockStore();
    it('should dispatch a fetchGamesListSuccess on success', () => {
      nock(endpoints.gameList)
        .get('/')
        .reply(
          200,
          JSON.stringify({
            data: mockGames
          })
        );

      store
        .dispatch(actions.fetchGamesList())
        .then(expect(store.getActions()).toMatchSnapshot());
    });

    it('should dispatch a failure action on error', () => {
      nock(endpoints.gameList)
        .get('/')
        .reply(500);

      store
        .dispatch(actions.fetchGamesList())
        .then(expect(store.getActions()).toMatchSnapshot());
    });
  });
});
