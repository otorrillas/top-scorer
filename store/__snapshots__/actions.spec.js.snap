// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Redux - actions FETCH_GAMES_LIST should dispatch a failure action on error 1`] = `Array []`;

exports[`Redux - actions FETCH_GAMES_LIST should dispatch a fetchGamesListSuccess on success 1`] = `Array []`;

exports[`Redux - actions export action types 1`] = `
Object {
  "FAILURE": "FAILURE",
  "FETCH_GAMES_LIST": "FETCH_GAMES_LIST",
  "FETCH_GAMES_LIST_SUCCESS": "FETCH_GAMES_LIST_SUCCESS",
  "FETCH_RUNS": "FETCH_RUNS",
  "FETCH_RUNS_SUCCESS": "FETCH_RUNS_SUCCESS",
  "RESET": "RESET",
}
`;

exports[`Redux - actions should create a FAILURE action 1`] = `
Object {
  "error": "test error",
  "type": "FAILURE",
}
`;

exports[`Redux - actions should create a FETCH_GAMES_LIST_SUCCESS action 1`] = `
Object {
  "data": Object {
    "1234": Object {
      "abbreviation": "test",
      "assets": Object {
        "cover-medium": Object {
          "uri": "https://cover-medium-url",
        },
      },
      "runs": Array [
        Object {
          "players": Array [
            Object {
              "id": "playerId",
            },
          ],
          "times": Object {
            "primary_t": 450,
          },
          "videos": Object {
            "links": Array [
              Object {
                "uri": "https://video-url",
              },
            ],
          },
        },
      ],
    },
  },
  "type": "FETCH_GAMES_LIST_SUCCESS",
}
`;

exports[`Redux - actions should create a FETCH_RUN_SUCCESS action 1`] = `
Object {
  "data": Array [
    Object {
      "players": Array [
        Object {
          "id": "playerId",
        },
      ],
      "times": Object {
        "primary_t": 450,
      },
      "videos": Object {
        "links": Array [
          Object {
            "uri": "https://video-url",
          },
        ],
      },
    },
  ],
  "id": 1234,
  "type": "FETCH_RUNS_SUCCESS",
}
`;

exports[`Redux - actions should create a RESET action 1`] = `
Object {
  "type": "RESET",
}
`;
